tests/basis_test.py

```diff
@@ -1,330 +1,438 @@
-
-def test_cheb_call_random_points():
-    """Test chebyshev polynomial at some degree at some input"""
-    numpy.random.seed(567)
-    ns = numpy.random.randint(20,size = 20)
-    xs = numpy.random.rand(20) * 2 - 1
-    for n,x in zip(ns,xs):
-        f = ChebyshevFirstKind(n)
-        assert numpy.isclose(f(x),special.eval_chebyt(n,x))
-
-def test_cheb_call_random_points_multi_input():
-    """Test that chebyshev polynomial call handles multiple x inputs"""
-    numpy.random.seed(567)
-    xs = numpy.random.rand(20) * 2 - 1
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    fn = ChebyshevFirstKind(5)
-    assert numpy.allclose(f0(xs),special.eval_chebyt(0,xs))
-    assert numpy.allclose(f1(xs),special.eval_chebyt(1,xs))
-    assert numpy.allclose(fn(xs),special.eval_chebyt(5,xs))
-
-
-def test_cheb_derivative():
-    """Test if the correct derivative is generated."""
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    f2 = ChebyshevFirstKind(2)
-    assert f0.derivative(1) == pytest.approx(0)
-    assert f1.derivative(1) == pytest.approx(1)
-    assert f2.derivative(1) == pytest.approx(4)
-    
-    assert f0.derivative(-0.5) == pytest.approx(0)
-    assert f1.derivative(-0.5) == pytest.approx(1)
-    assert f2.derivative(-0.5) == pytest.approx(-2)
-
-    assert numpy.isclose(f0.derivative([1,-0.5]),[0,0]).all()
-    assert numpy.isclose(f1.derivative([1,-0.5]),[1,1]).all()
-    assert numpy.isclose(f2.derivative([1,-0.5]),[4,-2]).all()
-
-
-def test_cheb_2nd_derivative():
-    """Test if the correct derivative is generated."""
-    u0 = ChebyshevSecondKind(0)
-    u1 = ChebyshevSecondKind(1)
-    u2 = ChebyshevSecondKind(2)
-    assert u0.derivative(1) == pytest.approx(0)
-    assert u1.derivative(1) == pytest.approx(2)
-    assert u2.derivative(1) == pytest.approx(8)
-
-    assert u0.derivative(-1) == pytest.approx(0)
-    assert u1.derivative(-1) == pytest.approx(2)
-    assert u2.derivative(-1) == pytest.approx(-8)
-
-    assert u0.derivative(0.5) == pytest.approx(0)
-    assert u1.derivative(0.5) == pytest.approx(2)
-    assert u2.derivative(0.5) == pytest.approx(4)
-
-    assert numpy.isclose(u0.derivative([1,-1,0.5]),[0,0,0]).all()
-    assert numpy.isclose(u1.derivative([1,-1,0.5]),[2,2,2]).all()
-    assert numpy.isclose(u2.derivative([1,-1,0.5]),[8,-8,4]).all()
-
-
-
-def test_cheb_call_invalid_input():
-    """Test call raises error if input is outside domain [-1,1]"""
-    f = ChebyshevFirstKind(4)
-    with pytest.raises(ValueError):
-        f(2)
-    with pytest.raises(ValueError):
-        f(-2)
-    with pytest.raises(ValueError):
-        f([0.5,0.7,3,0.8])
 
-
-def test_cheb_derivative_invalid_input():
-    """Test call raises error if input is outside domain [-1,1]"""
-    f = ChebyshevFirstKind(4)
-    with pytest.raises(ValueError):
-        f.derivative(2)
+import smolyay
+
+
+basis_call_answer_key = pytest.mark.parametrize(
+    "basis_fun,answer_key",
+    [
```

Comment:
For readability, I would suggest you just make this a list, and then use the `parametrize` decorator on the tests itself. That way it's clear what's going on, and also those tests get to pick what they want to call the argument names.
```suggestion
basis_call_answer_key =
    [
```
Please do the same for the others like this.